src/nvim/window.c

```diff
@@ -6145,6 +6167,110 @@ void set_fraction(win_T *wp)
   }
 }
 
+/*
+ * Handle scroll position for 'nosplitscroll', replaces
+ * scroll_to_fraction() call from win_new_height().
+ * Instead we iterate over all windows in a tabpage.
+ */
+static void spsc_correct_scroll(win_T *next_curwin, int flags)
+{
+  int state, curnormal, framewins, winmoved;
+  int tabwins = win_count();
+  long so;
+  frame_T *fr;
+  linenr_T lnum;
+
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab)
+  {
+    if (wp->w_floating || wp->w_botline < 3) {
+      goto skip;
+    }
+
+    lnum = wp->w_cursor.lnum;
+    // No need to correct scroll position if height has not changed.
+    // Might need to correct cursor when window was closed.
+    if (wp->w_prev_height == wp->w_height) {
+      if (flags & SPSC_CLOSE) {
+        spsc_correct_cursor(wp, lnum, FALSE);
+      }
+      goto skip;
+    }
+
+    winmoved = wp->w_winrow != wp->w_prev_winrow
+               && (wp->w_winrow > tabline_height() || (flags & SPSC_CLOSE));
+    // If winrow has moved, we maintain the same botline by setting
+    // cursor position, depending on 'laststatus' and number of
+    // windows in tab/frame. Scroll the cursor to FRACTION_MULT.
+    if (winmoved) {
+      if (wp->w_frame->fr_parent) {
+        framewins = 0;
+        FOR_ALL_FRAMES(fr, wp->w_frame->fr_parent->fr_child)
+        framewins++;
+      } else {
+        framewins = 1;
+      }
+      wp->w_cursor.lnum = wp->w_botline - ((p_ls == 1 && !(flags & SPSC_CLOSE)
+                                            && (framewins == 2 && tabwins == 2)) ? 2 : 1);
+    }
+
+    invalidate_botline_win(wp);
+    validate_botline(wp);
+    so = wp->w_p_so < 0 ? p_so : wp->w_p_so;
+    // When resizing shell, do not adjust botline.
+    if ((flags & SPSC_RESIZE)
+        && wp->w_cursor.lnum > (wp->w_botline - so - 1)) {
+      wp->w_cursor.lnum = MAX(1, (linenr_T)(wp->w_botline - so - 1));
+      goto skip;
+    } else if (winmoved) {
+      p_so = 0;
+      wp->w_fraction = FRACTION_MULT;
+      scroll_to_fraction(wp, wp->w_prev_height);
+      wp->w_cursor.lnum = lnum;
+      p_so = so;
+    }
+    // Ensure cursor position is valid for the current window to be
+    // or if currently not in normal/cmdline mode.
+    state = get_real_state();
+    curnormal = wp == curwin && state != MODE_NORMAL && state != MODE_CMDLINE;
+    if (wp == next_curwin || curnormal || winmoved) {
+      spsc_correct_cursor(wp, lnum, curnormal);
+    }
+skip:
+    wp->w_prev_winrow = wp->w_winrow;
+    wp->w_prev_height = wp->w_height;
+  }
+}
+
+/*
+ * Correct potentially invalid cursor position.
+ */
+static void spsc_correct_cursor(win_T *wp, linenr_T lnum, int curnormal)
```

Comment:
The docstring should explain how this is different from `fix_cursor`.

```suggestion
static void win_fix_cursor(win_T *wp, linenr_T lnum, int curnormal)
```